src/smc-webapp/course/common.cjsx

```diff
@@ -365,3 +368,180 @@ exports.StudentAssignmentInfo = rclass
+        @setState(show_selector : false)
+        @refs.search_input.clear_and_focus_search_input()
+
+    # TODO: Check if this works
+    search_button : ->
+        if @props.is_searching
+            # Currently doing a search, so show a spinner
+            <Button>
+                <Icon name="circle-o-notch" spin />
+            </Button>
+        else if @state.show_selector
+            # There is something in the selection box -- so only action is to clear the search box.
+            <Button onClick={@clear_and_focus_search_input}>
+                <Icon name="times-circle" />
+            </Button>
+        else
+            # Waiting for user to start a search
+            <Button onClick={@props.do_search}>
+                <Icon name="search" />
+            </Button>
+
+    add_button_clicked : (e) ->
+        e.preventDefault()
+        @props.add_selected(@state.selected_items)
+        @clear_and_focus_search_input()
+
+    render_results_list : ->
+        for item in @props.search_results
+            <option key={item} value={item} label={item}>{item}</option>
+
+    render_add_selector : ->
+        <div>
+            <Input type='select' multiple ref="selector" size=5 rows=10 onChange={=>@setState(selected_items : @refs.selector.getValue())}>
+                {@render_results_list()}
+            </Input>
+            <ButtonToolbar>
+                {@render_add_selector_button()}
+                <Button onClick={@clear_and_focus_search_input}>
+                    Cancel
+                </Button>
+            </ButtonToolbar>
+        </div>
+
+    render_add_selector_button : ->
+        num_items_selected = @state.selected_items?.length ? 0
+        btn_text = switch @props.search_results.length
+            when 0 then "No #{@props.item_name} found"
+            when 1 then "Add #{@props.item_name}"
+            else switch num_items_selected
+                when 0 then "Select #{@props.item_name} above"
+                when 1 then "Add selected #{@props.item_name}"
+                else "Add #{num_items_selected} #{@props.item_name}s"
+        disabled = @props.search_results.length == 0 or (@props.search_results.length >= 2 and num_items_selected == 0)
+        <Button disabled={disabled} onClick={@add_button_clicked}><Icon name="plus" /> {btn_text}</Button>
+
+    render : ->
+        <div>
+            <SearchInput
+                autoFocus     = {true}
+                ref           = 'search_input'
+                default_value = ''
+                placeholder   = "Add #{@props.item_name} by folder name (enter to see available folders)..."
+                on_submit     = {@props.do_search}
+                on_escape     = {@clear_and_focus_search_input}
+                buttonAfter   = {@search_button()}
+            />
+            {@render_add_selector() if @state.show_selector}
+         </div>
+
+# Definitely not a good abstraction.
+# Purely for code reuse (bad reason..)
```

Comment:
At least I think it's a bad reason for the very next comment. This "almost" pure enough result is why I didn't refactor out a lot of the other stuff that is *very* similar between `Assignment` and `Handout`.